File: packages/shared/src/pricing/guardrails.ts
// Deterministic guardrails — these wrap EVERY model output. No model bypasses them,
// which is what keeps the on-screen price reproducible and auditable: "why ₹1,050?" →
// point at the exact rule + feature values. The model proposes; these rules decide.
//
//   hard floor  = max(sellerFloor, routeElsewhereValue)  → below it, REROUTE (hand
//                 off to the Intelligent Bridge) instead of pricing lower
//   ceiling     = amazonNewPrice × 0.95                   → never within 5% of new
//   max step    = min(₹100, 8% of current price)          → no whiplash per reprice
//   rounding    = nearest ₹50                             → human-legible prices

import type { PricingStateVector, GuardrailResult } from './types.js';

export type GuardrailInput = {
  proposedPrice: number;
  currentPrice: number;
  state: PricingStateVector;
  isFirstListing: boolean;
};

export type GuardrailOutput = {
  finalPrice: number;
  /** true when the price hit the floor — signal the Bridge to reroute. */
  shouldReroute: boolean;
  guardrailsApplied: GuardrailResult[];
};

export function applyGuardrails(input: GuardrailInput): GuardrailOutput {
  const { proposedPrice, currentPrice, state, isFirstListing } = input;
  const applied: GuardrailResult[] = [];

  // 1. true floor and ceiling
  const floor = Math.max(state.sellerFloor, state.routeElsewhereValue);
  const ceiling = state.amazonNewPrice * 0.95; // never within 5% of new price

  let price = proposedPrice;

  // 2. hard floor — below this, reroute instead of pricing lower
  if (price < floor) {
    applied.push({ rule: 'hard_floor', triggered: true, adjustment: floor - price });
    return { finalPrice: floor, shouldReroute: true, guardrailsApplied: applied };
  }

  // 3. hard ceiling
  if (price > ceiling) {
    applied.push({ rule: 'ceiling', triggered: true, adjustment: ceiling - price });
    price = ceiling;
  }

  // 4. max change per step (skip on first listing — no prior price)
  if (!isFirstListing) {
    const maxStepChange = Math.min(100, currentPrice * 0.08); // ₹100 or 8%
    const delta = price - currentPrice;
    if (Math.abs(delta) > maxStepChange) {
      const clampedDelta = Math.sign(delta) * maxStepChange;
      applied.push({
        rule: 'max_step_change',
        triggered: true,
        adjustment: clampedDelta - delta,
      });
      price = currentPrice + clampedDelta;
    }
  }

  // 5. round to nearest ₹50
  const rounded = Math.round(price / 50) * 50;
  if (rounded !== price) {
    applied.push({ rule: 'round_to_50', triggered: true, adjustment: rounded - price });
    price = rounded;
  }

  // 6. final floor check after rounding
  if (price < floor) {
    price = floor;
    applied.push({ rule: 'post_round_floor', triggered: true });
  }

  return { finalPrice: price, shouldReroute: false, guardrailsApplied: applied };
}

File: ml/pricing/reloop_pricing/pricing/guardrails.py
"""Deterministic guardrails — 1:1 mirror of packages/shared/src/pricing/guardrails.ts.

The model proposes; these rules decide. Keeping a byte-for-byte-behaviour copy in Python
means the offline simulation clamps prices EXACTLY the way the live TS engine would, so a
price the agent learns from is the same price a buyer would have seen. The TS file is the
source of truth; change it there first, then update this mirror.

  hard floor = max(seller_floor, route_elsewhere_value)  → below it, REROUTE
  ceiling    = amazon_new_price × 0.95                    → never within 5% of new
  max step   = min(₹100, 8% of current)                  → no whiplash (skip first listing)
  rounding   = nearest ₹50                               → human-legible prices
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import List


@dataclass
class GuardrailResult:
    rule: str
    triggered: bool
    adjustment: float = 0.0


@dataclass
class GuardrailOutput:
    final_price: float
    should_reroute: bool  # True when the price hit the floor — signal the Bridge to reroute
    guardrails_applied: List[GuardrailResult]


def _round_half_up_50(price: float) -> float:
    """Nearest ₹50, rounding halves up (matches JS Math.round for positive prices)."""
    return math.floor(price / 50 + 0.5) * 50


def apply_guardrails(
    proposed_price: float,
    current_price: float,
    seller_floor: float,
    route_elsewhere_value: float,
    amazon_new_price: float,
    is_first_listing: bool,
) -> GuardrailOutput:
    applied: List[GuardrailResult] = []

    floor = max(seller_floor, route_elsewhere_value)
    ceiling = amazon_new_price * 0.95

    price = proposed_price

    # hard floor — below this, reroute instead of pricing lower
    if price < floor:
        applied.append(GuardrailResult("hard_floor", True, floor - price))
        return GuardrailOutput(floor, True, applied)

    # hard ceiling
    if price > ceiling:
        applied.append(GuardrailResult("ceiling", True, ceiling - price))
        price = ceiling

    # max change per step (skip on first listing — no prior price)
    if not is_first_listing:
        max_step = min(100.0, current_price * 0.08)
        delta = price - current_price
        if abs(delta) > max_step:
            clamped = math.copysign(max_step, delta)
            applied.append(GuardrailResult("max_step_change", True, clamped - delta))
            price = current_price + clamped

    # round to nearest ₹50
    rounded = _round_half_up_50(price)
    if rounded != price:
        applied.append(GuardrailResult("round_to_50", True, rounded - price))
        price = rounded

    # final floor check after rounding
    if price < floor:
        price = floor
        applied.append(GuardrailResult("post_round_floor", True))

    return GuardrailOutput(price, False, applied)
